File: sdk/cpp/examples/05_efficiency_curve.cpp
// examples/05_efficiency_curve.cpp -- upload an all-1.0 efficiency curve
// (a no-op identity correction), then commit it. Useful for verifying the
// upload path end-to-end without changing device behaviour.
//
// Usage: 05_efficiency_curve <serial-port> [count=711]
#include <cstdlib>
#include <iostream>
#include <string>
#include <vector>

#include "h1/Device.hpp"

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cerr << "usage: " << argv[0] << " <serial-port> [count=711]\n";
        return 2;
    }
    const std::size_t n = (argc >= 3) ? static_cast<std::size_t>(std::atoi(argv[2])) : 711;
    try {
        h1::Device dev(h1::openSerialPort(argv[1]));

        std::cout << "uploading identity efficiency curve (" << n << " floats)...\n";
        std::vector<float> ratios(n, 1.0f);
        dev.uploadEfficiencyCurve(ratios);
        std::cout << "verifying + committing...\n";
        dev.verifyAndComputeEfficiencyCurve();
        std::cout << "done. To revert to factory: call resetEfficiencyCurve().\n";
        return 0;
    } catch (const std::exception& e) {
        std::cerr << "error: " << e.what() << "\n";
        return 1;
    }
}

File: sdk/python/examples/05_efficiency_curve.py
"""Example 5: upload a flat unity efficiency curve, verify, then reset.

This is a destructive operation on a real device — it overwrites the
calibration. The script restores the factory default at the end.

Usage:
    python examples/05_efficiency_curve.py /dev/tty.usbserial-XYZ
"""

from __future__ import annotations

import sys

from h1_sdk import Device


def main(port: str) -> None:
    with Device(port) as dev:
        rng = dev.get_wavelength_range()
        # One ratio per spectrum point — set them all to 1.0 (identity).
        ratios = [1.0] * rng.count
        print(f"uploading {len(ratios)} ratios...")
        dev.upload_efficiency_curve(ratios)

        print("verifying and computing...")
        dev.verify_and_compute_efficiency_curve()
        print("OK")

        print("restoring factory curve...")
        dev.reset_efficiency_curve()
        print("OK")


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"usage: {sys.argv[0]} <serial-port>", file=sys.stderr)
        sys.exit(2)
    main(sys.argv[1])
